core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -109,9 +109,19 @@ case class LogReadResult(info: FetchDataInfo,
   def withEmptyFetchInfo: LogReadResult =
     copy(info = FetchDataInfo(LogOffsetMetadata.UnknownOffsetMetadata, MemoryRecords.EMPTY))
 
-  override def toString =
-    s"Fetch Data: [$info], HW: [$highWatermark], leaderLogStartOffset: [$leaderLogStartOffset], leaderLogEndOffset: [$leaderLogEndOffset], " +
-    s"followerLogStartOffset: [$followerLogStartOffset], fetchTimeMs: [$fetchTimeMs], readSize: [$readSize], lastStableOffset: [$lastStableOffset], error: [$error]"
+  override def toString = {
+    "LogReadResult(" +
+      s"info=$info, " +
+      s"highWatermark=$highWatermark, " +
+      s"leaderLogStartOffset=$leaderLogStartOffset, " +
+      s"leaderLogEndOffset=$leaderLogEndOffset, " +
+      s"followerLogStartOffset=$followerLogStartOffset, " +
+      s"fetchTimeMs=$fetchTimeMs, " +
+      s"preferredReadReplica=$preferredReadReplica, " +
+      s"lastStableOffset=$lastStableOffset, " +
+      s"error=$error" +
+      ")"
```

Comment:
Could we use the `toString` from the case class?

Reply:
Do we get the labels from the default `toString`? In the past, I thought it would only show the values.

Reply:
That's true. You have to write `toString` yourself if you want labels. We could write a utility method like so (didn't try too compile it):

```scala
public def productToString(product: Product): String = {
  val builder = new StringBuilder
  sb.append(product.prefix)
  for (i <- 0 until productArity) {
    builder.append(productElementName(i)).append("=").append(productElement(i))
  }
  sb.build()
}
```

Then we can call that method from any case class `toString` where we want this format. Avoids some duplication and forgetting to update `toString` when new methods are added.

Reply:
Ok, I added something like that.

Reply:
Sweet!